modules/mapper-extras/src/main/java/org/elasticsearch/index/mapper/RankFeatureFieldMapper.java

```diff
@@ -114,11 +115,11 @@ public Query existsQuery(QueryShardContext context) {
         }
 
         @Override
-        public ValueFetcher valueFetcher(MapperService mapperService, SearchLookup searchLookup, String format) {
+        public ValueFetcher valueFetcher(Supplier<Set<String>> sourcePaths, SearchLookup searchLookup, String format) {
```

Comment:
I wanted to "contain" MapperService so I've got a Supplier for the thing that you *might* need to get from it.

Reply:
To cut down on references to `MapperService`, we've been switching to suppliers or function references in a large number of places. To me it could make this code harder to read and follow. For example, now it's tricky to track down how the source paths are computed (you can't just ask for implementors of `Supplier` since there are so many!). It also feels like it's leaking some implementation details into the signature, because only `SourceValueFetcher` needs access to the source paths.

Reply:
I agree Julie, I am working on adding a new abstraction to QueryShardContext that allows to resolve field types and do mappings things. That can be carried around separately instead of passing multiple functions to different places. Would that address your concerns?

Reply:
I kind of thought of it the other way - that you need the MapperService at all leaks an implementation detail. I kind of liked that `SourceValueFetcher` only takes `Set<String>` now.